### ISSUE TITLE
D(D(f))(v) should work.

### ISSUE THREAD

**Issue by MasonProtter** [YingboMa/ForwardDiff2.jl#26]
Currently, if you want a Hessian you need to do `DI(DI(f))(v)` as `DI(D(f))(v)` errors out. Shouldn't `D(D(f))` just materialize the inner `D(f)`?

**Comment by YingboMa** [comment]
No, it shouldn't work. `DI` is made for this.

**Comment by shashi** [comment]
> `DI(D(f))(v)` errors out.

We should fix this.

@YingboMa Actually this makes sense right?? `D(f)` is literally `DI(f)` but happens not yet to be in a matrix or tensor form, but it is still a very nice mathematical function.

So `D(D(f))(v)` should work, regardless of whether we materialize the internal `D(f)` or not.

**Comment by shashi** [comment]
So it seems the basis for `D(f)` is not yet known so it doesn't make sense (@YingboMa said).

**Comment by YingboMa** [comment]
Yes, `D` doesn't mean "D(f)(x) * v computes df(x)/dx ⋅ v, taking advantage of the laziness in D(f)(x). However, if it is the only function that is inside another D, it actually means DI".

**Comment by MasonProtter** [comment]
The `I` in `DI` only exists to materialize the derivative. It only needs to be at the top level of the expression (if at all). 

> So it seems the basis for D(f) is not yet known so it doesn't make sense

Whatever basis information `D(f)` needs can be propagated to it. When you compute `D(D(f))(u) * v`, you have access to `u` and `v` before you descend into the `D(D(f))` calculation. There’s no missing information.

**Comment by YingboMa** [comment]
> The I in DI only exists to materialize the derivative.

`D(f)` doesn't mean to compute the derivative, gradient or Jacobian at all. `DI` is only a very special case of materializing it.

> Whatever basis information D(f) needs can be propagated to it. When you compute D(D(f))(u) * v, you have access to u and v before you descend into the D(D(f)) calculation. There’s no missing information.

There is. What should `D(D(prod))(ones(2)) * ones(2)` be? You really need to say `D(DI(prod))(ones(2)) * ones(2)`. You don't really want `D(x->D(prod)(x) * ones(2))(ones(2)) * ones(2)`, right?

```julia
julia> using ForwardDiff2: D, DI

julia> D(prod)(ones(2)) * ones(2)
1×1 LinearAlgebra.Adjoint{Float64,StaticArrays.SArray{Tuple{1},Float64,1,1}} with indices SOneTo(1)×SOneTo(1):
 2.0

julia> D(x->D(prod)(x) * ones(2))(ones(2)) * ones(2)
1-element StaticArrays.SArray{Tuple{1},Float64,1,1} with indices SOneTo(1):
 2.0

julia> D(DI(prod))(ones(2)) * ones(2)
2-element StaticArrays.SArray{Tuple{2},Float64,1,2} with indices SOneTo(2):
 1.0
 1.0
```
I am not trying to make the interface arbitrarily hard to use or anything. It is actually ambiguous.

**Comment by MasonProtter** [comment]
> D(f) doesn't mean to compute the derivative, gradient or Jacobian at all. DI is only a very special case of materializing it.

I think this is the crux of our disagreement. The way I see it, `D(f)` is the function which would produce the Jacobian, but since users don't always want the whole thing, we don't eagerly materialize it. For example, we know that if they do something like `D(f)(u) * v::Vector`, we won't need to allocate the full matrix and can instead just allocate a vector, so we punt on the computation until we know what sort of object the final output will go into. 

From you comments, I guess you don't see `D(f)` that way, but I think the API here encourages the sort of misconceptions I'm having. `I` is supposed to be the multiplicative identity. `X * I` should be **equal** to `X`. 

That leads me to think that if there are any differences between `D(f)(x)` and `D(f)(x) * I`, the only difference should be whether or not the computations in `D(f)(x)` have actually been run yet. Otherwise, I think your interface using things like `*` is actually a quite misleading pun.

**Comment by YingboMa** [comment]
Closed by #29
